File: packages/hg_native/android/src/main/kotlin/com/example/hg_native/HgNativePlugin.kt
package com.example.hg_native

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class HgNativePlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var store: SecureStoreEngine
    private val ai = LiteRtStubEngine()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        store = SecureStoreEngine(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, "hg_native")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "secureStore.write" -> {
                store.write(call.argument<String>("key")!!, call.argument<String>("value")!!)
                result.success(null)
            }
            "secureStore.read" -> result.success(store.read(call.argument<String>("key")!!))
            "secureStore.delete" -> {
                store.delete(call.argument<String>("key")!!)
                result.success(null)
            }
            "onDeviceAi.availability" -> result.success(ai.availability())
            "onDeviceAi.load" -> {
                ai.load()
                result.success(null)
            }
            "onDeviceAi.infer" -> result.success(
                ai.infer(
                    call.argument<String>("id") ?: "",
                    call.argument<String>("prompt") ?: "",
                ),
            )
            "onDeviceAi.cancel" -> result.success(null)
            "onDeviceAi.unload" -> result.success(null)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

/// v1: private prefs. Replace with EncryptedSharedPreferences / Keystore wrap.
private class SecureStoreEngine(context: Context) {
    private val prefs = context.getSharedPreferences("hg_native_secure", Context.MODE_PRIVATE)

    fun write(key: String, value: String) {
        prefs.edit().putString(key, value).apply()
    }

    fun read(key: String): String? = prefs.getString(key, null)

    fun delete(key: String) {
        prefs.edit().remove(key).apply()
    }
}

/** LiteRT / Gemini Nano adapter seam. Stub inference only. */
private class LiteRtStubEngine {
    fun availability(): Map<String, Any> = mapOf("ready" to true, "backend" to "litert-stub")

    fun load() {}

    fun infer(id: String, prompt: String): Map<String, Any> = mapOf(
        "requestId" to id,
        "text" to "On-device stub (LiteRT): $prompt",
        "done" to true,
    )
}

File: packages/hg_native/ios/Classes/HgNativePlugin.swift
import Flutter
import Security
import UIKit

public class HgNativePlugin: NSObject, FlutterPlugin {
  private let store = KeychainSecureStore()
  private let ai = CoreMlStubEngine()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "hg_native", binaryMessenger: registrar.messenger())
    let instance = HgNativePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    switch call.method {
    case "secureStore.write":
      store.write(key: args?["key"] as? String ?? "", value: args?["value"] as? String ?? "")
      result(nil)
    case "secureStore.read":
      result(store.read(key: args?["key"] as? String ?? ""))
    case "secureStore.delete":
      store.delete(key: args?["key"] as? String ?? "")
      result(nil)
    case "onDeviceAi.availability":
      result(ai.availability())
    case "onDeviceAi.load":
      result(nil)
    case "onDeviceAi.infer":
      result(ai.infer(id: args?["id"] as? String ?? "", prompt: args?["prompt"] as? String ?? ""))
    case "onDeviceAi.cancel", "onDeviceAi.unload":
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

private final class KeychainSecureStore {
  private let service = "hg_native"

  func write(key: String, value: String) {
    let data = Data(value.utf8)
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: key,
    ]
    SecItemDelete(query as CFDictionary)
    var add = query
    add[kSecValueData as String] = data
    SecItemAdd(add as CFDictionary, nil)
  }

  func read(key: String) -> String? {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: key,
      kSecReturnData as String: true,
      kSecMatchLimit as String: kSecMatchLimitOne,
    ]
    var item: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &item)
    guard status == errSecSuccess, let data = item as? Data else {
      return nil
    }
    return String(data: data, encoding: .utf8)
  }

  func delete(key: String) {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: service,
      kSecAttrAccount as String: key,
    ]
    SecItemDelete(query as CFDictionary)
  }
}

/// Core ML adapter seam. Stub inference only — no model weights.
private final class CoreMlStubEngine {
  func availability() -> [String: Any] {
    ["ready": true, "backend": "coreml-stub"]
  }

  func infer(id: String, prompt: String) -> [String: Any] {
    [
      "requestId": id,
      "text": "On-device stub (Core ML): \(prompt)",
      "done": true,
    ]
  }
}
